crypto_sign/tuov_ip/ref/tuov_keypair.c

```diff
@@ -0,0 +1,268 @@
+    memcpy(Q11, pk->pk + P12_BIAS, _PK_P1_BYTE >> 1);
+    memcpy(Q21, pk->pk + P22_BIAS, _PK_P2_BYTE >> 1);
+    memcpy(Q31, pk->pk + P32_BIAS, _PK_P3_BYTE >> 1);
+    multiple_S(Q123, sk->ST, N_TRIANGLE_TERMS(_V) + _V * _PUB_M);
+    gfv_add(Q11, pk->pk + P11_BIAS, _PK_P1_BYTE >> 1);
+    gfv_add(Q21, pk->pk + P21_BIAS, _PK_P2_BYTE >> 1);
+    gfv_add(Q31, pk->pk + P31_BIAS, _PK_P3_BYTE >> 1);
+
+    // line 8 - 12 of KeyGen
+    calculate_Q51(pk->pk + P51_BIAS, sk->ST + T1_BIAS, Q11, Q21);
+    calculate_Q61(pk->pk + P61_BIAS, sk->ST + T1_BIAS, sk->ST + T3_BIAS, sk->ST + T4_BIAS, Q11, Q21, Q31, pk->pk + P51_BIAS);
+    calculate_Q91(pk->pk + P91_BIAS, sk->ST + T3_BIAS, sk->ST + T4_BIAS, Q11, Q21, Q31, pk->pk + P51_BIAS, pk->pk + P61_BIAS);
+    calculate_Q91(pk->pk + P92_BIAS, sk->ST + T3_BIAS, sk->ST + T4_BIAS,
+                pk->pk + P12_BIAS, pk->pk + P22_BIAS, pk->pk + P32_BIAS,
+                pk->pk + P52_BIAS, pk->pk + P62_BIAS);
+
+    // Q569 stores Q52, Q62, Q92 in order
+    unsigned char Q569[(_PK_P5_BYTE + _PK_P6_BYTE + _PK_P9_BYTE) >> 1];
+    unsigned char *Q52 = Q569;
+    unsigned char *Q62 = Q569 + (_PK_P5_BYTE >> 1);
+    unsigned char *Q92 = Q569 + ((_PK_P5_BYTE + _PK_P6_BYTE) >> 1);
+    // line 13 of KeyGen
+    memcpy(Q52, pk->pk + P52_BIAS, _PK_P5_BYTE >> 1);
+    memcpy(Q62, pk->pk + P62_BIAS, _PK_P6_BYTE >> 1);
+    memcpy(Q92, pk->pk + P92_BIAS, _PK_P9_BYTE >> 1);
+    multiple_S(Q569, sk->ST, N_TRIANGLE_TERMS(_PUB_M));
+    gfv_add(pk->pk + P51_BIAS, Q52, _PK_P5_BYTE >> 1);
+    gfv_add(pk->pk + P61_BIAS, Q62, _PK_P6_BYTE >> 1);
+    gfv_add(pk->pk + P91_BIAS, Q92, _PK_P9_BYTE >> 1);
+
+    // now public key is done, we need to generate the F's in secret key
+
+    // F1 = Q1
+    memcpy(sk->F, Q123, _PK_P1_BYTE >> 1);
+    memcpy(sk->F + P12_BIAS, pk->pk + P12_BIAS, _PK_P1_BYTE >> 1);
+    // line 5-12 in ExpandSK
+    calculate_F21(sk->F + P21_BIAS, sk->ST + T1_BIAS, Q11, Q21);
+    calculate_F31(sk->F + P31_BIAS, sk->ST + T3_BIAS, sk->ST + T4_BIAS, Q11, Q21, Q31);
+    calculate_F21(sk->F + P22_BIAS, sk->ST + T1_BIAS, pk->pk + P12_BIAS, pk->pk + P22_BIAS);
+    calculate_F31(sk->F + P32_BIAS, sk->ST + T3_BIAS, sk->ST + T4_BIAS, pk->pk + P12_BIAS, pk->pk + P22_BIAS, pk->pk + P32_BIAS);
+    calculate_F52(sk->F + P51_BIAS, sk->ST + T1_BIAS, pk->pk + P12_BIAS, pk->pk + P22_BIAS, pk->pk + P52_BIAS);
+    calculate_F62(sk->F + P51_BIAS + (_PK_P5_BYTE >> 1), sk->ST + T1_BIAS, sk->ST + T3_BIAS, sk->ST + T4_BIAS,
+                    pk->pk + P12_BIAS, pk->pk + P22_BIAS, pk->pk + P32_BIAS, pk->pk + P52_BIAS, pk->pk + P62_BIAS);
+
+    return 0;
+}
+
+int generate_keypair(pk_t *pk, sk_t *sk, const unsigned char *pk_seed, const unsigned char *sk_seed){
+    int r = _generate_keypair(pk, sk, pk_seed, sk_seed);
+    combine_F(sk->F);
+    combine_P(pk->pk);
+    return r;
+}
+
+int generate_keypair_pkc( cpk_t *rpk, sk_t *sk, const unsigned char *pk_seed, const unsigned char *sk_seed ){
+    #if defined(_MALLOC_)
```

Comment:
This `malloc` is active for the larger parameter sets. Easily removed together with the `free` below... but might be just way too large for any stack.